File: android/src/main/kotlin/com/elriztechnology/unified_esc_pos_printer/BleManager.kt
package com.elriztechnology.unified_esc_pos_printer

import android.bluetooth.*
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import java.util.UUID

class BleManager(private val context: Context) {

    companion object {
        private val ESC_POS_SERVICE_UUID =
            UUID.fromString("000018f0-0000-1000-8000-00805f9b34fb")
        private val ESC_POS_TX_CHAR_UUID =
            UUID.fromString("00002af1-0000-1000-8000-00805f9b34fb")
    }

    private val bluetoothAdapter: BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter
    private val scanner: BluetoothLeScanner?
        get() = bluetoothAdapter?.bluetoothLeScanner

    private val mainHandler = Handler(Looper.getMainLooper())

    // Scan state (shared — only one scan at a time)
    private var scanEventSink: EventChannel.EventSink? = null
    private val discoveredDevices = mutableListOf<Map<String, String>>()
    private var scanCallback: ScanCallback? = null
    private var scanTimeoutRunnable: Runnable? = null

    // Per-device connection state
    private val gatts = mutableMapOf<String, BluetoothGatt>()
    private val txCharacteristics = mutableMapOf<String, BluetoothGattCharacteristic>()
    private val negotiatedMtus = mutableMapOf<String, Int>()
    private val writeWithoutResponses = mutableMapOf<String, Boolean>()
    private val connectResults = mutableMapOf<String, MethodChannel.Result>()
    private val writeResults = mutableMapOf<String, MethodChannel.Result>()
    private val targetServiceUuids = mutableMapOf<String, UUID>()
    private val targetCharUuids = mutableMapOf<String, UUID>()

    var connectionStateCallback: ((String, String) -> Unit)? = null

    val scanStreamHandler = object : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            scanEventSink = events
        }

        override fun onCancel(arguments: Any?) {
            scanEventSink = null
        }
    }

    fun getBondedBleDevices(result: MethodChannel.Result) {
        val adapter = bluetoothAdapter
        if (adapter == null) {
            result.success(emptyList<Map<String, String>>())
            return
        }

        try {
            val bonded = adapter.bondedDevices?.filter { device ->
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    device.type == BluetoothDevice.DEVICE_TYPE_LE ||
                            device.type == BluetoothDevice.DEVICE_TYPE_DUAL
                } else {
                    true // Include all if we can't check type
                }
            }?.map { device ->
                mapOf(
                    "deviceId" to device.address,
                    "name" to (device.name ?: device.address)
                )
            } ?: emptyList()
            result.success(bonded)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", "Cannot access bonded devices", e.message)
        }
    }

    fun startScan(timeoutMs: Long, result: MethodChannel.Result) {
        val s = scanner
        if (s == null) {
            result.error("UNAVAILABLE", "Bluetooth LE scanner not available", null)
            return
        }

        discoveredDevices.clear()

        scanCallback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, scanResult: ScanResult) {
                val device = scanResult.device
                val id = device.address
                if (discoveredDevices.none { it["deviceId"] == id }) {
                    val name = try { device.name } catch (_: SecurityException) { null }

                    discoveredDevices.add(
                        mapOf(
                            "deviceId" to id,
                            "name" to (name ?: id)
                        )
                    )

                    mainHandler.post {
                        scanEventSink?.success(discoveredDevices.toList())
                    }
                }
            }

            override fun onScanFailed(errorCode: Int) {
                mainHandler.post {
                    scanEventSink?.error("SCAN_FAILED", "BLE scan failed with code $errorCode", null)
                }
            }
        }

        try {
            s.startScan(scanCallback)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", "Bluetooth scan permission denied", e.message)
            return
        }

        // Auto-stop after timeout
        scanTimeoutRunnable = Runnable {
            stopScanInternal()
        }

        mainHandler.postDelayed(scanTimeoutRunnable!!, timeoutMs)

        result.success(null)
    }

    fun stopScan(result: MethodChannel.Result) {
        stopScanInternal()
        result.success(null)
    }

    private fun stopScanInternal() {
        scanTimeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        scanTimeoutRunnable = null
        scanCallback?.let { cb ->
            try {
                scanner?.stopScan(cb)
            } catch (_: SecurityException) {
                // Already lost permission — ignore
            }
        }
        scanCallback = null
    }

    fun connect(
        deviceId: String,
        timeoutMs: Long,
        serviceUuid: String?,
        characteristicUuid: String?,
        result: MethodChannel.Result
    ) {
        val adapter = bluetoothAdapter
        if (adapter == null) {
            result.error("UNAVAILABLE", "Bluetooth adapter not available", null)
            return
        }

        targetServiceUuids[deviceId] = serviceUuid?.let { UUID.fromString(it) } ?: ESC_POS_SERVICE_UUID
        targetCharUuids[deviceId] = characteristicUuid?.let { UUID.fromString(it) } ?: ESC_POS_TX_CHAR_UUID

        val device: BluetoothDevice
        try {
            device = adapter.getRemoteDevice(deviceId)
        } catch (e: Exception) {
            result.error("INVALID_DEVICE", "Invalid device ID: $deviceId", e.message)
            return
        }

        // Clean up any existing connection for this device
        cleanupConnection(deviceId)

        connectResults[deviceId] = result

        // Timeout handler
        val timeoutRunnable = Runnable {
            val pendingResult = connectResults.remove(deviceId)
            if (pendingResult != null) {
                pendingResult.error("TIMEOUT", "BLE connection timed out", null)

                try { gatts[deviceId]?.disconnect(); gatts[deviceId]?.close() } catch (_: SecurityException) {}
                gatts.remove(deviceId)
            }
        }
        mainHandler.postDelayed(timeoutRunnable, timeoutMs)

        val gattCallback = object : BluetoothGattCallback() {
            override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
                if (newState == BluetoothProfile.STATE_CONNECTED) {
                    gatts[deviceId] = g
                    try {
                        g.requestMtu(512)
                    } catch (e: SecurityException) {
                        mainHandler.post {
                            mainHandler.removeCallbacks(timeoutRunnable)
                            connectResults.remove(deviceId)?.error("PERMISSION_DENIED", "MTU request denied", e.message)
                        }
                    }
                } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                    mainHandler.post {
                        mainHandler.removeCallbacks(timeoutRunnable)

                        val pendingResult = connectResults.remove(deviceId)
                        if (pendingResult != null) {
                            pendingResult.error("DISCONNECTED", "BLE device disconnected during setup", null)
                        } else {
                            // Remote disconnection after fully connected
                            connectionStateCallback?.invoke(deviceId, "disconnected")
                        }

                        try { g.close() } catch (_: SecurityException) {}

                        gatts.remove(deviceId)
                        txCharacteristics.remove(deviceId)
                    }
                }
            }

            override fun onMtuChanged(g: BluetoothGatt, mtu: Int, status: Int) {
                negotiatedMtus[deviceId] = if (status == BluetoothGatt.GATT_SUCCESS) mtu - 3 else 20
                try {
                    g.discoverServices()
                } catch (e: SecurityException) {
                    mainHandler.post {
                        mainHandler.removeCallbacks(timeoutRunnable)
                        connectResults.remove(deviceId)?.error("PERMISSION_DENIED", "Service discovery denied", e.message)
                    }
                }
            }

            override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    mainHandler.post {
                        mainHandler.removeCallbacks(timeoutRunnable)
                        connectResults.remove(deviceId)?.error("SERVICE_DISCOVERY_FAILED", "GATT service discovery failed with status $status", null)
                        try { g.disconnect(); g.close() } catch (_: SecurityException) {}
                        gatts.remove(deviceId)
                    }
                    return
                }

                var foundChar: BluetoothGattCharacteristic? = null

                // 1. Try target service/characteristic UUIDs
                val targetService = g.getService(targetServiceUuids[deviceId])
                if (targetService != null) {
                    val c = targetService.getCharacteristic(targetCharUuids[deviceId])
                    if (c != null && isWritable(c)) {
                        foundChar = c
                    }
                }

                // 2. Fallback: any writable characteristic
                if (foundChar == null) {
                    for (service in g.services) {
                        for (c in service.characteristics) {
                            if (isWritable(c)) {
                                foundChar = c
                                break
                            }
                        }
                        if (foundChar != null) break
                    }
                }

                mainHandler.post {
                    mainHandler.removeCallbacks(timeoutRunnable)
                    if (foundChar == null) {
                        connectResults.remove(deviceId)?.error("NO_CHARACTERISTIC", "No writable characteristic found", null)
                        try { g.disconnect(); g.close() } catch (_: SecurityException) {}
                        gatts.remove(deviceId)
                    } else {
                        txCharacteristics[deviceId] = foundChar
                        // Prefer write-with-response for reliable backpressure; the printer
                        // ACKs each chunk before we send the next, preventing buffer overflow.
                        // Fall back to write-without-response only if that is the sole option.
                        writeWithoutResponses[deviceId] =
                            (foundChar.properties and BluetoothGattCharacteristic.PROPERTY_WRITE) == 0 &&
                            (foundChar.properties and BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE) != 0
                        connectResults.remove(deviceId)?.success(null)
                        connectionStateCallback?.invoke(deviceId, "connected")
                    }
                }
            }

            override fun onCharacteristicWrite(
                g: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
                status: Int
            ) {
                mainHandler.post {
                    val pendingResult = writeResults.remove(deviceId)
                    if (status == BluetoothGatt.GATT_SUCCESS) {
                        pendingResult?.success(null)
                    } else {
                        pendingResult?.error("WRITE_FAILED", "BLE write failed with status $status", null)
                    }
                }
            }
        }

        // Check if device is bonded — use autoConnect=true for bonded devices
        // as they may not be advertising, but the system can connect when they
        // become available.
        val isBonded = try {
            device.bondState == BluetoothDevice.BOND_BONDED
        } catch (_: SecurityException) {
            false
        }

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                device.connectGatt(context, isBonded, gattCallback, BluetoothDevice.TRANSPORT_LE)
            } else {
                device.connectGatt(context, isBonded, gattCallback)
            }
        } catch (e: SecurityException) {
            mainHandler.removeCallbacks(timeoutRunnable)
            connectResults.remove(deviceId)
            result.error("PERMISSION_DENIED", "Bluetooth connect permission denied", e.message)
        }
    }

    fun getMtu(deviceId: String, result: MethodChannel.Result) {
        result.success(negotiatedMtus[deviceId] ?: 20)
    }

    fun supportsWriteWithoutResponse(deviceId: String, result: MethodChannel.Result) {
        result.success(writeWithoutResponses[deviceId] ?: false)
    }

    fun write(deviceId: String, data: ByteArray, withoutResponse: Boolean, result: MethodChannel.Result) {
        val g = gatts[deviceId]
        val char = txCharacteristics[deviceId]
        if (g == null || char == null) {
            result.error("NOT_CONNECTED", "BLE device not connected: $deviceId", null)
            return
        }

        writeResults[deviceId] = result

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                // Android 13+ uses new writeCharacteristic API
                val writeType = if (withoutResponse)
                    BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
                else
                    BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT

                    val writeResult = g.writeCharacteristic(char, data, writeType)
                if (writeResult != BluetoothStatusCodes.SUCCESS) {
                    writeResults.remove(deviceId)?.error("WRITE_FAILED", "writeCharacteristic returned $writeResult", null)
                }
            } else {
                @Suppress("DEPRECATION")
                char.writeType = if (withoutResponse)
                    BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
                else
                    BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT

                @Suppress("DEPRECATION")
                char.value = data

                @Suppress("DEPRECATION")
                val success = g.writeCharacteristic(char)
                if (!success) {
                    writeResults.remove(deviceId)?.error("WRITE_FAILED", "writeCharacteristic returned false", null)
                }
            }
        } catch (e: SecurityException) {
            writeResults.remove(deviceId)?.error("PERMISSION_DENIED", "Bluetooth write permission denied", e.message)
        }
    }

    fun disconnect(deviceId: String, result: MethodChannel.Result) {
        cleanupConnection(deviceId)
        connectionStateCallback?.invoke(deviceId, "disconnected")
        result.success(null)
    }

    fun dispose() {
        stopScanInternal()

        val deviceIds = gatts.keys.toList()
        for (deviceId in deviceIds) {
            cleanupConnection(deviceId)
        }
    }

    private fun cleanupConnection(deviceId: String) {
        try {
            gatts[deviceId]?.disconnect()
            gatts[deviceId]?.close()
        } catch (_: SecurityException) {}

        gatts.remove(deviceId)
        txCharacteristics.remove(deviceId)
        negotiatedMtus.remove(deviceId)
        writeWithoutResponses.remove(deviceId)
        connectResults.remove(deviceId)
        writeResults.remove(deviceId)
        targetServiceUuids.remove(deviceId)
        targetCharUuids.remove(deviceId)
    }

    private fun isWritable(c: BluetoothGattCharacteristic): Boolean {
        val props = c.properties
        return (props and BluetoothGattCharacteristic.PROPERTY_WRITE) != 0 ||
                (props and BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE) != 0
    }
}

File: ios/Classes/BleManager.swift
import CoreBluetooth
import Flutter

class BleManager: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

    private static let escPosServiceUUID = CBUUID(string: "000018f0-0000-1000-8000-00805f9b34fb")
    private static let escPosTxCharUUID = CBUUID(string: "00002af1-0000-1000-8000-00805f9b34fb")

    private var centralManager: CBCentralManager?
    private var scanEventSink: FlutterEventSink?
    private var discoveredDevices: [[String: String]] = []
    private var scanTimer: Timer?

    // Per-device connection state, keyed by peripheral UUID string
    private var connectedPeripherals: [String: CBPeripheral] = [:]
    private var txCharacteristics: [String: CBCharacteristic] = [:]
    private var notifyCharacteristics: [String: CBCharacteristic] = [:]
    private var mtuPayloads: [String: Int] = [:]
    private var canWriteWithoutResponses: [String: Bool] = [:]
    private var connectResults: [String: FlutterResult] = [:]
    private var writeResults: [String: FlutterResult] = [:]
    private var pendingWriteDatas: [String: Data] = [:]
    private var pendingWriteResults: [String: FlutterResult] = [:]
    private var connectTimers: [String: Timer] = [:]
    private var writeTimeoutWorkItems: [String: DispatchWorkItem] = [:]
    private var targetServiceUUIDs: [String: CBUUID] = [:]
    private var targetCharUUIDs: [String: CBUUID] = [:]

    /// Backstop timeout for a single BLE write. CoreBluetooth resolves a
    /// write only via `didWriteValueFor` (with-response) or
    /// `peripheralIsReady` (without-response, buffer-full path); if the
    /// printer stalls or vanishes without ever firing `didDisconnectPeripheral`,
    /// neither callback arrives and the FlutterResult would leak — hanging
    /// the Dart-side `await bleWrite` forever. This bounds that wait.
    private let writeTimeoutSeconds: TimeInterval = 8.0

    var connectionStateCallback: ((String, String) -> Void)?
    var incomingBytesCallback: ((String, Data) -> Void)?

    /// Continuations waiting for CBCentralManager to leave the `.unknown` state.
    /// Each is called with `true` when the state becomes `.poweredOn`, or
    /// `false` on any other resolved state / timeout.
    private var poweredOnContinuations: [(Bool) -> Void] = []

    // Lazy init to avoid triggering the BT permission dialog on plugin load
    private func ensureCentralManager() {
        if centralManager == nil {
            centralManager = CBCentralManager(delegate: self, queue: nil)
        }
    }

    /// Waits for the CBCentralManager to reach `.poweredOn`.
    ///
    /// - If already `.poweredOn`, calls `completion(true)` synchronously.
    /// - If the state is `.unknown` (adapter still initialising), queues the
    ///   completion and resolves it once `centralManagerDidUpdateState` fires.
    /// - Falls back to `completion(false)` after [timeout] seconds.
    private func waitForPoweredOn(timeout: TimeInterval = 5.0,
                                  completion: @escaping (Bool) -> Void) {
        guard let cm = centralManager else {
            completion(false)
            return
        }

        if cm.state == .poweredOn {
            completion(true)
            return
        }

        // Already resolved to a non-poweredOn state — fail immediately.
        if cm.state != .unknown {
            completion(false)
            return
        }

        // State is .unknown — manager is still initialising, wait for update.
        var resolved = false
        let resolve: (Bool) -> Void = { isPoweredOn in
            guard !resolved else { return }
            resolved = true
            completion(isPoweredOn)
        }

        poweredOnContinuations.append(resolve)

        DispatchQueue.main.asyncAfter(deadline: .now() + timeout) {
            resolve(false)
        }
    }

    lazy var scanStreamHandler: FlutterStreamHandler = {
        return BleScanStreamHandler(manager: self)
    }()

    func setScanEventSink(_ sink: FlutterEventSink?) {
        scanEventSink = sink
    }

    func requestPermissions(result: @escaping FlutterResult) {
        ensureCentralManager()
        if #available(iOS 13.1, *) {
            switch CBCentralManager.authorization {
            case .allowedAlways:
                result(true)
            case .notDetermined:
                result(true)
            case .denied, .restricted:
                result(false)
            @unknown default:
                result(true)
            }
        } else {
            result(centralManager?.state == .poweredOn)
        }
    }

    func getBondedBleDevices(result: @escaping FlutterResult) {
        ensureCentralManager()
        guard let cm = centralManager, cm.state == .poweredOn else {
            result([[[String: String]]]())
            return
        }

        let connected = cm.retrieveConnectedPeripherals(withServices: [
            BleManager.escPosServiceUUID
        ])

        let devices: [[String: String]] = connected.map { peripheral in
            [
                "deviceId": peripheral.identifier.uuidString,
                "name": peripheral.name ?? peripheral.identifier.uuidString
            ]
        }
        result(devices)
    }

    func startScan(timeoutMs: Int, result: @escaping FlutterResult) {
        ensureCentralManager()

        waitForPoweredOn { [weak self] isPoweredOn in
            guard let self = self, let cm = self.centralManager, isPoweredOn else {
                result(FlutterError(code: "UNAVAILABLE", message: "Bluetooth is not powered on", details: nil))
                return
            }

            self.discoveredDevices.removeAll()
            cm.scanForPeripherals(withServices: nil, options: [
                CBCentralManagerScanOptionAllowDuplicatesKey: false
            ])

            let timeout = Double(timeoutMs) / 1000.0
            self.scanTimer = Timer.scheduledTimer(withTimeInterval: timeout, repeats: false) { [weak self] _ in
                self?.stopScanInternal()
            }

            result(nil)
        }
    }

    func stopScan(result: @escaping FlutterResult) {
        stopScanInternal()
        result(nil)
    }

    private func stopScanInternal() {
        scanTimer?.invalidate()
        scanTimer = nil
        centralManager?.stopScan()
        // Signal Flutter that no more scan events are coming. Closes the
        // Dart-side broadcast stream gracefully so a subsequent
        // subscription.cancel() becomes a no-op instead of producing
        // "No active stream to cancel" noise in services-library logs
        // (and Crashlytics).
        DispatchQueue.main.async {
            self.scanEventSink?(FlutterEndOfEventStream)
            self.scanEventSink = nil
        }
    }

    func connect(deviceId: String, timeoutMs: Int, serviceUuid: String?, characteristicUuid: String?, result: @escaping FlutterResult) {
        ensureCentralManager()

        waitForPoweredOn { [weak self] isPoweredOn in
            guard let self = self, let cm = self.centralManager, isPoweredOn else {
                result(FlutterError(code: "UNAVAILABLE", message: "Bluetooth is not powered on", details: nil))
                return
            }

            guard let uuid = UUID(uuidString: deviceId) else {
                result(FlutterError(code: "INVALID_DEVICE", message: "Invalid device UUID: \(deviceId)", details: nil))
                return
            }

            self.targetServiceUUIDs[deviceId] = serviceUuid != nil ? CBUUID(string: serviceUuid!) : BleManager.escPosServiceUUID
            self.targetCharUUIDs[deviceId] = characteristicUuid != nil ? CBUUID(string: characteristicUuid!) : BleManager.escPosTxCharUUID

            let peripherals = cm.retrievePeripherals(withIdentifiers: [uuid])
            guard let peripheral = peripherals.first else {
                result(FlutterError(code: "NOT_FOUND", message: "Peripheral not found for UUID: \(deviceId)", details: nil))
                return
            }

            // Clean up any existing connection
            self.cleanupConnection(deviceId)

            self.connectResults[deviceId] = result
            self.connectedPeripherals[deviceId] = peripheral
            peripheral.delegate = self

            cm.connect(peripheral, options: nil)

            // Timeout
            let timeout = Double(timeoutMs) / 1000.0
            self.connectTimers[deviceId] = Timer.scheduledTimer(withTimeInterval: timeout, repeats: false) { [weak self] _ in
                guard let self = self, self.connectResults[deviceId] != nil else { return }
                cm.cancelPeripheralConnection(peripheral)
                self.connectResults.removeValue(forKey: deviceId)?(FlutterError(code: "TIMEOUT", message: "BLE connection timed out", details: nil))
                self.connectedPeripherals.removeValue(forKey: deviceId)
            }
        }
    }

    func getMtu(deviceId: String, result: @escaping FlutterResult) {
        result(mtuPayloads[deviceId] ?? 20)
    }

    func supportsWriteWithoutResponse(deviceId: String, result: @escaping FlutterResult) {
        result(canWriteWithoutResponses[deviceId] ?? false)
    }

    func write(deviceId: String, data: Data, withoutResponse: Bool, result: @escaping FlutterResult) {
        guard let peripheral = connectedPeripherals[deviceId], let char = txCharacteristics[deviceId] else {
            result(FlutterError(code: "NOT_CONNECTED", message: "BLE device not connected: \(deviceId)", details: nil))
            return
        }

        // Fast path: buffer has room, the write is fire-and-forget, so it
        // can never stall waiting on a callback — resolve immediately.
        if withoutResponse && peripheral.canSendWriteWithoutResponse {
            peripheral.writeValue(data, for: char, type: .withoutResponse)
            result(nil)
            return
        }

        // One-shot guard so neither the CoreBluetooth callback, the
        // backstop timeout, nor a disconnect cleanup can resolve the same
        // FlutterResult twice.
        var settled = false
        let settle: FlutterResult = { [weak self] value in
            guard !settled else { return }
            settled = true
            self?.writeTimeoutWorkItems.removeValue(forKey: deviceId)?.cancel()
            result(value)
        }

        if withoutResponse {
            pendingWriteDatas[deviceId] = data
            pendingWriteResults[deviceId] = settle
        } else {
            writeResults[deviceId] = settle
            peripheral.writeValue(data, for: char, type: .withResponse)
        }

        // Backstop: fail the write if no callback arrives in time so the
        // Dart-side await can't hang forever on a silently stalled link.
        let workItem = DispatchWorkItem { [weak self] in
            guard let self = self else { return }
            self.writeResults.removeValue(forKey: deviceId)
            self.pendingWriteResults.removeValue(forKey: deviceId)
            self.pendingWriteDatas.removeValue(forKey: deviceId)
            settle(FlutterError(code: "WRITE_FAILED", message: "BLE write timed out", details: nil))
        }
        writeTimeoutWorkItems[deviceId] = workItem
        DispatchQueue.main.asyncAfter(deadline: .now() + writeTimeoutSeconds, execute: workItem)
    }

    /// Sends DLE EOT status query via write-with-response.  The write ACK
    /// confirms the printer's GATT server received the data.
    func queryStatus(deviceId: String, timeoutMs: Int, result: @escaping FlutterResult) {
        guard let peripheral = connectedPeripherals[deviceId], let char = txCharacteristics[deviceId] else {
            result(FlutterError(code: "NOT_CONNECTED", message: "BLE device not connected: \(deviceId)", details: nil))
            return
        }

        let dleEot = Data([0x10, 0x04, 0x01])
        var responded = false

        // didWriteValueFor will call writeResults[deviceId] which we set here.
        writeResults[deviceId] = { [weak self] writeResult in
            guard !responded else { return }
            responded = true
            self?.writeResults.removeValue(forKey: deviceId)
            if let error = writeResult as? FlutterError {
                result(-1)
            } else {
                result(0)
            }
        }
        peripheral.writeValue(dleEot, for: char, type: .withResponse)

        // Timeout fallback
        let timeoutSec = Double(timeoutMs) / 1000.0
        DispatchQueue.main.asyncAfter(deadline: .now() + timeoutSec) { [weak self] in
            guard !responded else { return }
            responded = true
            self?.writeResults.removeValue(forKey: deviceId)
            result(-1)
        }
    }

    func disconnect(deviceId: String, result: @escaping FlutterResult) {
        if let peripheral = connectedPeripherals[deviceId], let cm = centralManager {
            cm.cancelPeripheralConnection(peripheral)
        }
        cleanupConnection(deviceId)
        connectionStateCallback?(deviceId, "disconnected")
        result(nil)
    }

    private func cleanupConnection(_ deviceId: String) {
        connectTimers.removeValue(forKey: deviceId)?.invalidate()
        writeTimeoutWorkItems.removeValue(forKey: deviceId)?.cancel()
        // Fail any in-flight write so the Dart-side `await bleWrite`
        // throws immediately on disconnect instead of leaking the
        // FlutterResult and hanging the print job forever. The settle
        // guard makes a double-resolution (callback already fired) a
        // no-op, so this is safe to call unconditionally.
        let disconnectError = FlutterError(code: "DISCONNECTED", message: "BLE device disconnected during write", details: nil)
        writeResults.removeValue(forKey: deviceId)?(disconnectError)
        pendingWriteResults.removeValue(forKey: deviceId)?(disconnectError)
        // Disable notifications on the cached notify characteristic, if
        // any, before dropping the reference. iOS Core Bluetooth keeps
        // a subscription on the peripheral even after we let go of the
        // characteristic object — leaving stale notifications running
        // would feed bytes to nobody and confuse the next reconnect.
        if let peripheral = connectedPeripherals[deviceId],
           let nf = notifyCharacteristics[deviceId],
           peripheral.state == .connected {
            peripheral.setNotifyValue(false, for: nf)
        }
        connectedPeripherals.removeValue(forKey: deviceId)
        txCharacteristics.removeValue(forKey: deviceId)
        notifyCharacteristics.removeValue(forKey: deviceId)
        mtuPayloads.removeValue(forKey: deviceId)
        canWriteWithoutResponses.removeValue(forKey: deviceId)
        pendingWriteDatas.removeValue(forKey: deviceId)
        pendingWriteResults.removeValue(forKey: deviceId)
        connectResults.removeValue(forKey: deviceId)
        writeResults.removeValue(forKey: deviceId)
        targetServiceUUIDs.removeValue(forKey: deviceId)
        targetCharUUIDs.removeValue(forKey: deviceId)
    }

    /// Resolve the deviceId (peripheral UUID string) for a given CBPeripheral.
    private func deviceId(for peripheral: CBPeripheral) -> String {
        return peripheral.identifier.uuidString
    }

    // MARK: - CBCentralManagerDelegate

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Resolve any pending "wait for powered on" continuations.
        if central.state != .unknown {
            let isPoweredOn = central.state == .poweredOn
            let pending = poweredOnContinuations
            poweredOnContinuations.removeAll()
            for continuation in pending {
                continuation(isPoweredOn)
            }
        }
        // If BT turns off, ongoing connections will trigger
        // didDisconnectPeripheral automatically.
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral,
                         advertisementData: [String: Any], rssi RSSI: NSNumber) {
        let id = peripheral.identifier.uuidString
        if !discoveredDevices.contains(where: { $0["deviceId"] == id }) {
            discoveredDevices.append([
                "deviceId": id,
                "name": peripheral.name ?? id
            ])
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }
                self.scanEventSink?(self.discoveredDevices)
            }
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        peripheral.discoverServices(nil)
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        let id = deviceId(for: peripheral)
        connectTimers.removeValue(forKey: id)?.invalidate()
        connectResults.removeValue(forKey: id)?(FlutterError(code: "CONNECTION_FAILED", message: "Failed to connect: \(error?.localizedDescription ?? "unknown")", details: nil))
        connectedPeripherals.removeValue(forKey: id)
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        let id = deviceId(for: peripheral)
        if connectResults[id] != nil {
            connectTimers.removeValue(forKey: id)?.invalidate()
            connectResults.removeValue(forKey: id)?(FlutterError(code: "DISCONNECTED", message: "Disconnected during setup", details: nil))
        } else {
            // Remote disconnection
            connectionStateCallback?(id, "disconnected")
        }
        cleanupConnection(id)
    }

    // MARK: - CBPeripheralDelegate

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        let id = deviceId(for: peripheral)
        if let error = error {
            connectTimers.removeValue(forKey: id)?.invalidate()
            centralManager?.cancelPeripheralConnection(peripheral)
            connectResults.removeValue(forKey: id)?(FlutterError(code: "SERVICE_DISCOVERY_FAILED", message: error.localizedDescription, details: nil))
            return
        }

        if let services = peripheral.services {
            for service in services {
                peripheral.discoverCharacteristics(nil, for: service)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        let id = deviceId(for: peripheral)
        if error != nil { return }

        guard txCharacteristics[id] == nil else { return } // Already found

        var txChar: CBCharacteristic?
        var notifyChar: CBCharacteristic?

        // 1. Try target service/characteristic UUIDs first.
        if let targetSvc = targetServiceUUIDs[id], service.uuid == targetSvc,
           let chars = service.characteristics {
            for c in chars {
                if txChar == nil && c.uuid == targetCharUUIDs[id] && isWritable(c) {
                    txChar = c
                }
                if notifyChar == nil && isNotifiable(c) {
                    notifyChar = c
                }
            }
        }

        // 2. Fallback: any writable characteristic on this service, plus
        //    any notifiable sibling for ASB / status pushes.
        if txChar == nil, let chars = service.characteristics {
            for c in chars {
                if txChar == nil && isWritable(c) {
                    txChar = c
                }
                if notifyChar == nil && isNotifiable(c) {
                    notifyChar = c
                }
            }
        }

        guard let tx = txChar else { return }

        if let nf = notifyChar {
            notifyCharacteristics[id] = nf
            peripheral.setNotifyValue(true, for: nf)
        }

        selectCharacteristic(tx, peripheral: peripheral)
    }

    private func selectCharacteristic(_ char: CBCharacteristic, peripheral: CBPeripheral) {
        let id = deviceId(for: peripheral)
        txCharacteristics[id] = char
        // Prefer write-without-response whenever the characteristic supports
        // it, even if it also advertises write-with-response. Most ESC/POS
        // printers expose both; with-response forces a per-chunk ATT ACK
        // round-trip (and a prepared/long-write sequence for chunks > MTU),
        // sending only one packet per connection event — an order of
        // magnitude slower. Without-response lets iOS pipeline several
        // packets per event with flow control via
        // peripheralIsReady(toSendWriteWithoutResponse:).
        let useWithoutResponse = char.properties.contains(.writeWithoutResponse)
        canWriteWithoutResponses[id] = useWithoutResponse

        let writeType: CBCharacteristicWriteType = useWithoutResponse ? .withoutResponse : .withResponse
        mtuPayloads[id] = peripheral.maximumWriteValueLength(for: writeType)

        connectTimers.removeValue(forKey: id)?.invalidate()
        connectResults.removeValue(forKey: id)?(nil)
        connectionStateCallback?(id, "connected")
    }

    /// Forwards bytes from a subscribed notify / indicate characteristic to
    /// the Dart side via [incomingBytesCallback]. Consumed by the per-
    /// connector AsbMonitor to parse Automatic Status Back (ESC/POS
    /// `GS a`) packets.
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if error != nil { return }
        guard let value = characteristic.value, !value.isEmpty else { return }
        let id = deviceId(for: peripheral)
        // Copy the value to break any aliasing with the underlying Core
        // Bluetooth buffer before we hop to the main queue.
        let payload = Data(value)
        if let cb = incomingBytesCallback {
            DispatchQueue.main.async {
                cb(id, payload)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        let id = deviceId(for: peripheral)
        if let error = error {
            writeResults.removeValue(forKey: id)?(FlutterError(code: "WRITE_FAILED", message: error.localizedDescription, details: nil))
        } else {
            writeResults.removeValue(forKey: id)?(nil)
        }
    }

    func peripheralIsReady(toSendWriteWithoutResponse peripheral: CBPeripheral) {
        let id = deviceId(for: peripheral)
        guard let data = pendingWriteDatas.removeValue(forKey: id),
              let char = txCharacteristics[id] else { return }
        let result = pendingWriteResults.removeValue(forKey: id)
        peripheral.writeValue(data, for: char, type: .withoutResponse)
        result?(nil)
    }

    private func isWritable(_ c: CBCharacteristic) -> Bool {
        return c.properties.contains(.write) || c.properties.contains(.writeWithoutResponse)
    }

    private func isNotifiable(_ c: CBCharacteristic) -> Bool {
        return c.properties.contains(.notify) || c.properties.contains(.indicate)
    }
}

class BleScanStreamHandler: NSObject, FlutterStreamHandler {
    private weak var manager: BleManager?

    init(manager: BleManager) {
        self.manager = manager
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        manager?.setScanEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        manager?.setScanEventSink(nil)
        return nil
    }
}
